### PR TITLE
#86 Build plugin and images for multiple architectures

### LINKED ISSUE #86: Build docker images for arm64 architecture
We would like to use this plugin on arm instances. It works well on amd64.
Would it be possible to do a multi arch build and publish of the container images?

### PR DESCRIPTION
**Pull request by larhauga**
This adds support for building and publishing linux-amd64 and linux-arm64 versions of the plugin.

This fixes #86

See https://github.com/larhauga/opa-docker-authz/actions for example run and https://github.com/larhauga/opa-docker-authz/pkgs/container/opa-docker-authz for published package

I have only gotten this to work by explicitly pinning to the tag `edge-linux-arm64`/`edge-linux-amd64` and I think we can live with this solution in the interim.

---
**Details regarding multi-arch manifest list**

When trying to build a manifest list for a multi-arch tag, it does not work, and we get the following error that seems to be coming from here: https://github.com/moby/moby/blob/master/plugin/backend_linux.go#L203

```
msg="Handler for GET /v1.44/plugins/privileges returned error: did not find plugin config for specified reference ghcr.io/larhauga/opa-docker-authz:edge"
Error response from daemon: did not find plugin config for specified reference ghcr.io/larhauga/opa-docker-authz:edge
```

I have created a Q&A at the moby repo to get some input on this: https://github.com/moby/moby/discussions/47369

The manifests created with `docker manifest` does not create a manifest usable by the docker plugin system. For future references a custom manifest can be created in the following way:
```bash
curl -X PUT -H "Authorization: Bearer $(echo $GITHUB_TOKEN|base64)" -H "Content-Type: application/vnd.docker.distribution.manifest.v2+json" --data @manifest.json https://ghcr.io/v2/${{ github.repository }}/manifests/edge
```

This is the `edge-linux-amd64` manifest.
```json
{
        "schemaVersion": 2,
        "config": {
                "mediaType": "application/vnd.docker.plugin.v1+json",
                "digest": "sha256:f588fb1f8e38f38b53d49c1b79281b36c2ca64b2b16119c1c8c5b6019af7d6c6",
                "size": 836
        },
        "layers": [
                {
                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                        "digest": "sha256:18eeb142bdc5e4d0b518e1b1321fa0a4863b21248ba614e07d4dcd87966e3d31",
                        "size": 13257330
                }
        ],
        "mediaType": "application/vnd.docker.distribution.manifest.v2+json"
}
```

The `:edge` manifest:
```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 419,
         "digest": "sha256:9b40bdeec15886807829ef61be7483bbc556758176944426965220cc2afe0f52",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 419,
         "digest": "sha256:aac32815a31f558589e17ed28deeb500c482d557d9038710f4e84177941d10a1",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}

```

### CODE REVIEW

File: .github/workflows/pull-request.yaml [hunk 1]
```diff
@@ -10,7 +10,11 @@ jobs:
     steps:
     - name: Check out code
       uses: actions/checkout@v3
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v3
 
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
```
Comment (anderseknert) [inline .github/workflows/pull-request.yaml:17]: It looks like these steps are unused in this workflow, or do they do something?

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -31,4 +37,16 @@ jobs:
...
+
+          # docker currently does not support multi-arch plugins
+          #docker manifest create ${{ env.REPO }}:${{ env.TAG_NAME }} \
+          #  ${{ env.REPO }}:${{ env.TAG_NAME }}-linux-amd64 \
+          #  ${{ env.REPO }}:${{ env.TAG_NAME }}-linux-arm64
+          
+          #docker manifest annotate  ${{ env.REPO }}:${{ env.TAG_NAME }} ${{ env.REPO }}:${{ env.TAG_NAME }}-linux-amd64 --os linux --arch amd64
+          #docker manifest annotate  ${{ env.REPO }}:${{ env.TAG_NAME }} ${{ env.REPO }}:${{ env.TAG_NAME }}-linux-arm64 --os linux --arch arm64
+
+          #docker manifest push ${{ env.REPO }}:${{ env.TAG_NAME }}
```
Comment (anderseknert) [inline .github/workflows/main.yaml:52]: Unless this is something that'll work soon, I'd get rid of the commented out code here and in release.yaml

**Review by anderseknert — COMMENTED** [review]
Thanks! Just a few nits, but looks good 👍

File: .github/workflows/pull-request.yaml (hunk 1 above)
Comment (larhauga) [reply .github/workflows/pull-request.yaml:17]: Both these actions are needed for building multiple architectures without running on seperate runners that has specific architectures. 
The QEMU action setups tooling needed to emulate different architectures used by docker buildx.
And we use buildx in the `plugin.sh` build step to specify which platform should be used.

Comment (anderseknert) [reply .github/workflows/pull-request.yaml:17]: Gotcha! Thanks 👍

**Review by anderseknert — APPROVED** [review]
Thanks! Let's give it a try :)

**Pull request merged**